adoc/extensions/sycl_khr_free_function_commands.adoc

```diff
@@ -0,0 +1,928 @@
+= SYCL_KHR_FREE_FUNCTION_COMMANDS
+
+This extension provides an alternative mechanism for submitting commands to a
+device via free-functions that require developers to opt-in to the creation of
+[code]#event# objects.
+
+== Dependencies
+
+This extension has no dependencies on other extensions.
+
+== Feature test macro
+
+An implementation supporting this extension must predefine the macro
+[code]#SYCL_KHR_FREE_FUNCTION_COMMANDS# to one of the values defined in the
+table below.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+== Usage example
+
+The example below rewrites the application from
+<<subsec:example.sycl.application>> to demonstrate the usage of this extension.
+
+[source,role=synopsis]
+----
+#include <iostream>
+#include <sycl/sycl.hpp>
+using namespace sycl; // (optional) avoids need for "sycl::" before SYCL names.
+
+// Size of the matrices.
+constexpr size_t N = 2000;
+constexpr size_t M = 3000;
+
+int main() {
+  // Create a queue to work on.
+  queue myQueue;
+
+  // Create some 2D arrays of float for our matrices.
+  float* a = malloc_shared<float>(N * M, myQueue);
+  float* b = malloc_shared<float>(N * M, myQueue);
+  float* c = malloc_shared<float>(N * M, myQueue);
+
+  // Launch an asynchronous kernel to initialize a.
+  // Use khr::submit to enqueue a kernel via a handler.
+  khr::submit(myQueue, [&](handler& cgh) {
+
+    // Enqueue a kernel iterating on a N*M 2D iteration space.
+    khr::launch(cgh, range<2> { N, M }, [=](id<2> index) {
+      size_t i = index[0];
+      size_t j = index[1];
+      a[i * M + j] = i * 2 + j;
+    });
+
+  });
+
+  // Launch an asynchronous kernel to initialize b.
+  // Use khr::launch to enqueue a kernel without a handler.
+  khr::launch(myQueue, range<2> { N, M }, [=](id<2> index) {
+    size_t i = index[0];
+    size_t j = index[1];
+    b[i * M + j] = i * 2014 + j * 42;
+  });
+
+  // Ensure that the two previous kernels complete before enqueueing more work.
+  // This does not block the host, but enforces dependencies on the device.
+  khr::command_barrier(myQueue);
+
+  // Launch an asynchronous kernel to compute matrix addition c = a + b.
+  // Use khr::launch_grouped to enqueue a kernel using groups without a handler.
+  range<2> local = { 2, 2 };
+  range<2> global = { N, M };
+  khr::launch_grouped(myQueue, global, local, [=](khr::invocation<2> ivc) {
```

Comment:
Is the `invocation` type part of this KHR?  If so, I think it is underspecified.  The description of `launch_grouped` says that it is equivalent to `h.parallel_for(nd_range(r, size), k)`.  However, that's not true if kernel's call operator is expected to take an `invocation` instead of an `nd_item`.

Reply:
Ah, I think I was missing a constructor, which I've added in 16111b2d8a9e5ccd685d61a905dce9f158d094e8.

This way, `khr::launch_grouped` expects a lambda taking anything implicitly constructible from a `sycl::nd_item` (which is equivalent to the behavior of `parallel_for`).  The implementation automatically calls the `khr::invocation` constructor when it launches the kernel.

Reply:
It seems like `khr::invocation` is orthogonal to the rest of this extension, then?  As you say, `handler::parallel_for` already allows the kernel parameter to be any type that is convertible from `nd_item`, so `khr::invocation` is also usable in kernels launched that way.

Why should `khr::invocation` be part of this extension rather than its own extension?  Or, maybe it should be part of "sycl_khr_group_interface"?  Is the main purpose of `khr::invocation` to align the naming with that extension?

Reply:
Good question...

I don't think there's any dependency here that would prevent us from splitting things out into a separate extension if that's what we want to do.  But if functionality _can_ be meaningfully grouped together, it might be better to have fewer extensions -- that means fewer macros for developers to check, and less chance of a perception that we introduced "too many extensions".

If I recall correctly, we added `khr::invocation` here because we were discussing the free-function commands when it was proposed.  `khr::launch_grouped` removes the need for any `nd_range` object, and somebody suggested that we could use that as an opportunity to remove `nd_item` too.  So, the proposal for the SYCL-Next interface is a `launch_grouped` accepting an `invocation` -- that would be my main justification for putting them in the same extension.

You're right that the naming is aligned with "sycl_khr_group_interface", but I think that's just a consequence of trying to make sure all these different concepts would be compatible if we choose to promote all the KHRs to SYCL-Next.

Reply:
> If I recall correctly, we added `khr::invocation` here because we were discussing the free-function commands when it was proposed. `khr::launch_grouped` removes the need for any `nd_range` object, and somebody suggested that we could use that as an opportunity to remove `nd_item` too.

I understand the motivation for removing `nd_range` from the API.  There was no way to do list-initialization of an `nd_range`, so you couldn't do something like `launch_grouped({SIZE},{WGSIZE}}, ...)`.  Splitting it into two `range` parameters solves this, because you *can* do `launch_grouped({SIZE}, {WGSIZE}, ...)`.

I don't understand the motivation for renaming `nd_item` to `invocation`.  On the surface, this just seems like superfluous naming churn.  If the main motivation is just to clarify the naming, it seems like this should be part of some larger extension.  For example, if we don't like the name `nd_item`, do we still like the name `item`?  It seems weird to rename the one and not the other.

Reply:
As discussed offline, I'm convinced that we should continue discussion about potential replacements for `nd_item` elsewhere.  Prototyping and testing the rest of this KHR does not rely on a replacement for `nd_item`, so I've removed the `khr::invocation` class in 269706cd5c2255b77204cabdb75d6595a068baa2.